File: android/src/main/java/com/theoplayer/ads/AdsModule.java
package com.theoplayer.ads;

import android.util.Log;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.theoplayer.SourceHelper;
import com.theoplayer.android.api.ads.dai.GoogleDaiIntegration;
import com.theoplayer.android.api.error.THEOplayerException;
import com.theoplayer.util.ViewResolver;

public class AdsModule extends ReactContextBaseJavaModule {
  private static final String TAG = "AdsModule";
  private final SourceHelper sourceHelper = new SourceHelper();
  private final ViewResolver viewResolver;

  public AdsModule(ReactApplicationContext context) {
    super(context);
    viewResolver = new ViewResolver(context);
  }

  @NonNull
  @Override
  public String getName() {
    return TAG;
  }

  // Add an ad break request.
  @ReactMethod
  public void schedule(Integer tag, ReadableMap ad) {
    viewResolver.resolveViewByTag(tag, view -> {
      if (view != null) {
        try {
          view.getAdsApi().schedule(sourceHelper.parseAdFromJS(ad));
        } catch (THEOplayerException exception) {
          Log.e(TAG, exception.getMessage());
        }
      }
    });
  }

  // The currently playing ad break.
  @ReactMethod
  public void currentAdBreak(Integer tag, Promise promise) {
    viewResolver.resolveViewByTag(tag, view -> {
      if (view == null) {
        promise.resolve(Arguments.createMap());
      } else {
        promise.resolve(AdInfo.fromAdBreak(view.getAdsApi().getCurrentAdBreak()));
      }
    });
  }

  // List of currently playing ads.
  @ReactMethod
  public void currentAds(Integer tag, Promise promise) {
    viewResolver.resolveViewByTag(tag, view -> {
      if (view == null) {
        promise.resolve(Arguments.createMap());
      } else {
        promise.resolve(AdInfo.fromAds(view.getAdsApi().getCurrentAds()));
      }
    });
  }

  // List of ad breaks which still need to be played.
  @ReactMethod
  public void scheduledAdBreaks(Integer tag, Promise promise) {
    viewResolver.resolveViewByTag(tag, view -> {
      if (view == null) {
        promise.resolve(Arguments.createMap());
      } else {
        promise.resolve(AdInfo.fromAdBreaks(view.getAdsApi().getScheduledAdBreaks()));
      }
    });
  }

  // Whether a linear ad is currently playing.
  @ReactMethod
  public void playing(Integer tag, Promise promise) {
    viewResolver.resolveViewByTag(tag, view -> {
      if (view == null) {
        promise.resolve(false);
      } else {
        promise.resolve(view.getAdsApi().isPlaying());
      }
    });
  }

  // Skip the current linear ad.
  // NOTE: This will have no effect when the current linear ad is (not yet) skippable.
  @ReactMethod
  public void skip(Integer tag) {
    viewResolver.resolveViewByTag(tag, view -> {
      if (view != null) {
        view.getAdsApi().skip();
      }
    });
  }

  @ReactMethod
  public void daiSnapback(Integer tag, Promise promise) {
    viewResolver.resolveViewByTag(tag, view -> {
      GoogleDaiIntegration daiIntegration = view != null ? view.getDaiIntegration() : null;
      if (daiIntegration == null) {
        promise.resolve(false);
      } else {
        promise.resolve(daiIntegration.getEnableSnapback());
      }
    });
  }

  @ReactMethod
  public void daiSetSnapback(Integer tag, Boolean enabled) {
    viewResolver.resolveViewByTag(tag, view -> {
      GoogleDaiIntegration daiIntegration = view != null ? view.getDaiIntegration() : null;
      if (daiIntegration != null) {
        daiIntegration.setEnableSnapback(enabled);
      }
    });
  }

  @ReactMethod
  public void daiContentTimeForStreamTime(Integer tag, Integer time, Promise promise) {
    viewResolver.resolveViewByTag(tag, view -> {
      GoogleDaiIntegration daiIntegration = view != null ? view.getDaiIntegration() : null;
      if (daiIntegration == null) {
        promise.resolve(time);
      } else {
        promise.resolve((int)(1e03 * daiIntegration.contentTimeForStreamTime(1e-03 * time)));
      }
    });
  }

  @ReactMethod
  public void daiStreamTimeForContentTime(Integer tag, Integer time, Promise promise) {
    viewResolver.resolveViewByTag(tag, view -> {
      GoogleDaiIntegration daiIntegration = view != null ? view.getDaiIntegration() : null;
      if (daiIntegration == null) {
        promise.resolve(time);
      } else {
        promise.resolve((int)(1e03 * daiIntegration.streamTimeForContentTime(1e-03 * time)));
      }
    });
  }
}

File: android/src/main/java/com/theoplayer/ads/AdsModule.kt
package com.theoplayer.ads

import android.util.Log
import android.view.View
import com.facebook.react.bridge.*
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.module.model.ReactModuleInfo
import com.theoplayer.source.SourceAdapter
import com.theoplayer.util.ViewResolver
import com.theoplayer.ReactTHEOplayerView
import com.theoplayer.android.api.ads.OmidFriendlyObstruction
import com.theoplayer.android.api.ads.OmidFriendlyObstructionPurpose
import com.theoplayer.android.api.error.THEOplayerException

private const val PROP_OMID_VIEW = "view"
private const val PROP_OMID_PURPOSE = "purpose"
private const val PROP_OMID_REASON = "reason"

private const val ERR_SCHEDULE_AD = "Error scheduling ad"

@Suppress("unused")
@ReactModule(name = AdsModule.NAME)
class AdsModule(context: ReactApplicationContext) : ReactContextBaseJavaModule(context) {
  companion object {
    const val NAME = "THEORCTAdsModule"
    val INFO = ReactModuleInfo(
      name = NAME,
      className = NAME,
      canOverrideExistingModule = false,
      needsEagerInit = false,
      isCxxModule = false,
      isTurboModule = false,
    )
  }

  private val sourceHelper = SourceAdapter()
  private val viewResolver: ViewResolver = ViewResolver(context)

  override fun getName(): String {
    return NAME
  }

  // Add an ad break request.
  @ReactMethod
  fun schedule(tag: Int, ad: ReadableMap) {
    viewResolver.resolveViewByTag(tag) { view: ReactTHEOplayerView? ->
      try {
        view?.adsApi?.schedule(sourceHelper.parseAdFromJS(ad))
      } catch (exception: THEOplayerException) {
        Log.e(NAME, exception.message ?: ERR_SCHEDULE_AD)
      }
    }
  }

  // The currently playing ad break.
  @ReactMethod
  fun currentAdBreak(tag: Int, promise: Promise) {
    viewResolver.resolveViewByTag(tag) { view: ReactTHEOplayerView? ->
      promise.resolve(if (view == null) Arguments.createMap() else AdAdapter.fromAdBreak(view.adsApi.currentAdBreak))
    }
  }

  // List of currently playing ads.
  @ReactMethod
  fun currentAds(tag: Int, promise: Promise) {
    viewResolver.resolveViewByTag(tag) { view: ReactTHEOplayerView? ->
      promise.resolve(if (view == null) Arguments.createArray() else AdAdapter.fromAds(view.adsApi.currentAds))
    }
  }

  // List of ad breaks which still need to be played.
  @ReactMethod
  fun scheduledAdBreaks(tag: Int, promise: Promise) {
    viewResolver.resolveViewByTag(tag) { view: ReactTHEOplayerView? ->
      promise.resolve(if (view == null) Arguments.createArray() else AdAdapter.fromAdBreaks(view.adsApi.scheduledAdBreaks))
    }
  }

  // Whether a linear ad is currently playing.
  @ReactMethod
  fun playing(tag: Int, promise: Promise) {
    viewResolver.resolveViewByTag(tag) { view: ReactTHEOplayerView? ->
      promise.resolve(view?.adsApi?.isPlaying ?: false)
    }
  }

  // Skip the current linear ad.
  // NOTE: This will have no effect when the current linear ad is not (yet) skippable.
  @ReactMethod
  fun skip(tag: Int) {
    viewResolver.resolveViewByTag(tag) { view: ReactTHEOplayerView? -> view?.adsApi?.skip() }
  }

  @ReactMethod
  fun daiSnapback(tag: Int, promise: Promise) {
    viewResolver.resolveViewByTag(tag) { view: ReactTHEOplayerView? ->
      promise.resolve(view?.playerContext?.daiIntegration?.enableSnapback ?: false)
    }
  }

  @ReactMethod
  fun daiSetSnapback(tag: Int, enabled: Boolean?) {
    viewResolver.resolveViewByTag(tag) { view: ReactTHEOplayerView? ->
      view?.playerContext?.daiIntegration?.let { integration ->
        integration.enableSnapback = enabled ?: false
      }
    }
  }

  @ReactMethod
  fun daiContentTimeForStreamTime(tag: Int, time: Int, promise: Promise) {
    viewResolver.resolveViewByTag(tag) { view: ReactTHEOplayerView? ->
      val contentTime = view?.playerContext?.daiIntegration?.let { integration ->
        (1e03 * integration.contentTimeForStreamTime(1e-03 * time)).toInt()
      } ?: time
      promise.resolve(contentTime)
    }
  }

  @ReactMethod
  fun daiStreamTimeForContentTime(tag: Int, time: Int, promise: Promise) {
    viewResolver.resolveViewByTag(tag) { view: ReactTHEOplayerView? ->
      val streamTime = view?.playerContext?.daiIntegration?.let { integration ->
        (1e03 * integration.streamTimeForContentTime(1e-03 * time)).toInt()
      } ?: time
      promise.resolve(streamTime)
    }
  }

  @ReactMethod
  fun addFriendlyObstruction(tag: Int, obstruction: ReadableMap) {
    val obsTag =
      if (obstruction.hasKey(PROP_OMID_VIEW)) obstruction.getInt(PROP_OMID_VIEW) else null
    val purpose = if (obstruction.hasKey(PROP_OMID_PURPOSE)) {
      when (obstruction.getString(PROP_OMID_PURPOSE)) {
        "videoControls" -> OmidFriendlyObstructionPurpose.VIDEO_CONTROLS
        "closeAd" -> OmidFriendlyObstructionPurpose.CLOSE_AD
        "notVisible" -> OmidFriendlyObstructionPurpose.NOT_VISIBLE
        else -> OmidFriendlyObstructionPurpose.OTHER
      }
    } else null
    val reason =
      if (obstruction.hasKey(PROP_OMID_REASON)) obstruction.getString(PROP_OMID_REASON) else null

    if (obsTag !== null) {
      viewResolver.resolveViewByTag(obsTag) { obsView: View? ->
        addFriendlyObstruction(tag, obsView, purpose, reason)
      }
    }
  }

  private fun addFriendlyObstruction(
    tag: Int,
    obsView: View?,
    purpose: OmidFriendlyObstructionPurpose?,
    reason: String?
  ) {
    if (obsView == null || purpose == null) return
    viewResolver.resolveViewByTag(tag) { view: ReactTHEOplayerView? ->
      view?.player?.ads?.omid?.addFriendlyObstruction(
        OmidFriendlyObstruction(obsView, purpose, reason)
      )
    }
  }

  @ReactMethod
  fun removeAllFriendlyObstructions(tag: Int) {
    viewResolver.resolveViewByTag(tag) { view: ReactTHEOplayerView? ->
      view?.player?.ads?.omid?.removeAllFriendlyObstructions()
    }
  }
}
